src/procedures/proc_shortest_path.c

```diff
@@ -0,0 +1,291 @@
+/*
+* Copyright 2018-2020 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "proc_shortest_path.h"
+#include "proc_ctx.h"
+#include "../query_ctx.h"
+#include "../datatypes/array.h"
+#include "../datatypes/path/path.h"
+#include "../graph/entities/node.h"
+#include "../algorithms/LAGraph_BF_full.h"
+
+// CALL algo.shortestPath(startNode, endNode, weightProperty, relationshipQuery, defaultValue)
+
+typedef struct {
+	bool pathProduced;              // Indicates if a path was produced.
+	Node *startNode;                // Start node to construct path from.
+	Node *endNode;                  // End node, path ends with this node.
+	const char *weightProperty;     // Attribute holding edge weight.
+	const char *relationship;       // Type of relationship to traverse.
+	double defaultValue;            // Use default value for edges missing weight attribute.
+	GrB_Matrix M;                   // Weights matrix.
+	SIValue *output;                // Output ["Path", path, "Cost", cost].
+} ShortestPath_Ctx;
+
+/* Construct NXN matrix M
+ * where M's main diagonal M[i,i] = 0
+ * for every edge of type `relation` connecting node i to node j
+ * W[i,j] = weight_attr of edge. */
+static GrB_Matrix _BuildWeightMatrix(const char *relation, const char *weight_attr,
+									 double default_weight) {
+
+	assert(relation && weight_attr);
+	GrB_Matrix M = GrB_NULL;
+
+	GraphContext *gc = QueryCtx_GetGraphCtx();
+	Schema *s = GraphContext_GetSchema(gc, relation, SCHEMA_EDGE);
+	if(!s) return GrB_NULL; // Relationship type doesn't exists.
+
+	Attribute_ID attr_id = GraphContext_GetAttributeID(gc, weight_attr);
+	if(attr_id == ATTRIBUTE_NOTFOUND) return GrB_NULL;  // Attribute doesn't exists.
+
+	Graph *g = QueryCtx_GetGraph();
+	GrB_Index n = Graph_RequiredMatrixDim(g);
+	GrB_Matrix R = Graph_GetRelationMatrix(g, s->id);
+
+	// TODO: enable FP64 in GB_control.h
+	GrB_Info info = GrB_Matrix_new(&M, GrB_FP64, n, n);
+	if(info != GrB_SUCCESS) return GrB_NULL;
+
+	// Extract relations out of matrix.
+	GrB_Index nz;
+	GrB_Matrix_nvals(&nz, R);
+	GrB_Index *I = rm_malloc(sizeof(GrB_Index) * nz);
+	GrB_Index *J = rm_malloc(sizeof(GrB_Index) * nz);
+	uint64_t  *X = rm_malloc(sizeof(uint64_t) * nz);
+	info = GrB_Matrix_extractTuples_UINT64(I, J, X, &nz, R);
+
+	// Process each relation.
+	for(GrB_Index i = 0; i < nz; i++) {
```

Comment:
use open mp?

Reply:
Can't `GrB_Matrix_setElement_FP64` isn't thread safe.
I don't believe we can call this function from multiple threads on the same matrix concurrently.